lib/ansible/modules/network/nxos/nxos_bfd_global.py

```diff
@@ -0,0 +1,308 @@
+from ansible.module_utils.network.common.config import CustomNetworkConfig
+from ansible.module_utils.connection import ConnectionError
+
+BFD_CMD_REF = '''
+# BFD does not support json format yet; use cli config strings instead.
+---
+echo_interface:
+  type: str
+  getval: (no )*bfd echo-interface *(\S+)*$
+  setval: '{0}bfd echo-interface {1}'
+  default: ''
+
+echo_rx_interval:
+  type: int
+  getval: bfd echo-rx-interval (\d+)$
+  setval: bfd echo-rx-interval {0}
+  default: 50
+  N3K: 250
+
+interval:
+  type: int_list
+  getval: bfd interval (\d+) min_rx (\d+) multiplier (\d+)
+  setval: bfd interval {0} min_rx {1} multiplier {2}
+  default: [50,50,3]
+  N3K: [250,250,3]
+
+slow_timer:
+  type: int
+  getval: bfd slow-timer (\d+)$
+  setval: bfd slow-timer {0}
+  default: 2000
+'''
+BFD_SHOW_CMD = "show run bfd all | incl '^(no )*bfd'"
+
+
+def init_cmd_ref(module):
+    '''Ensure BFD feature is enabled. Create cmd_ref dict.
+    '''
+    output = execute_show_command(module, 'show run bfd', 'text')
+    if output and 'CLI command error' in output:
+        if module.check_mode:
+            msg = "** 'feature bfd' is disabled. BFD will auto-enable when not in check_mode **"
+            module.warn(msg)
+        else:
+            load_config(module, 'feature bfd')
+
+    cmd_ref = yaml.load(BFD_CMD_REF)
+    cmd_ref = get_platform_defaults(module, cmd_ref)
+    return cmd_ref
+
+def get_platform_defaults(module, cmd_ref):
+    '''Get platform type and update cmd_ref with platform specific defaults
+    '''
+    cap = get_capabilities(module)
+    if cap:
+        device_info = cap['device_info']['network_os_platform']
+        plat = None
+        if 'N3K' in device_info:
+            plat = 'N3K'
+        if plat:
+            for k in cmd_ref:
+                if plat in cmd_ref[k]:
+                    cmd_ref[k]['default'] = cmd_ref[k][plat]
+
+    return cmd_ref
+
+
+def execute_show_command(module, command, format):
+    cmds = [{
+        'command': command,
+        'output': format,
+    }]
+    output = None
+    try:
+        output = run_commands(module, cmds)[0]
+    except ConnectionError as exc:
+        if 'CLI command error' in repr(exc):
+            output = repr(exc)
+        else:
+            raise
+    return output
+
+
+def get_existing(module, cmd_ref, show_cmd):
+    '''
+    Get a list of existing command states from device; then update cmd_ref
+    with any 'existing' values that differ from default states
+    '''
+    output = execute_show_command(module, show_cmd, 'text') or []
+    if not output:
+        return {}
+
+    # Walk each cmd in cmd_ref, use cmd pattern to discover existing cmds
+    output = output.split('\n')
+    for k in cmd_ref:
+        pattern = cmd_ref[k]['getval']
+        match = [m.groups() for m in (re.search(pattern, line) for line in output) if m]
+        if not match:
+            continue
+        match = list(match[0]) # tuple to list
```

Comment:
So if you had multiple lines that matched, this would discard all but the first match? Is that intended/desirable, or should multiple matches be considered an error to report?

Reply:
I went with the simple approach since there are no multiple instances in the BFD case. If the ansible folks go along with this implementation then we can add multiple instance support later.

For now I'll check for multiple instances and raise if that occurs.